Cesium3DTilesSelection/src/WebMapServiceRasterOverlay.cpp

```diff
@@ -0,0 +1,410 @@
+    try {
+      const int maxWidth = std::stoi(maxWidthText);
+      const int optionalTileWidth = options.tileWidth.value_or(256);
+      if (optionalTileWidth > maxWidth) {
+        char buffer[512];
+        std::sprintf(
+            buffer,
+            "configured tile width (%d) exceeds "
+            "Service >> "
+            "MaxWidth defined in WMS document (%d).",
+            optionalTileWidth,
+            maxWidth);
+        error = buffer;
+        return false;
+      }
+    } catch (std::invalid_argument&) { 
+      error = "Invalid web map service XML document";
+      return false;
+    }
+  }
+
+  tinyxml2::XMLElement* pOptionalServiceMaxHeight =
+      pService->FirstChildElement("MaxHeight");
+  if (pOptionalServiceMaxHeight) {
+    const std::string maxHeightText = pOptionalServiceMaxHeight->GetText();
+    try {
+      const int maxHeight = std::stoi(maxHeightText);
+      const int optionalTileHeight = options.tileHeight.value_or(256);
+      if (optionalTileHeight > maxHeight) {
+        char buffer[512];
+        std::sprintf(
+            buffer,
+            "configured tile height (%d) exceeds "
+            "Service >> "
+            "MaxHeight defined in WMS document (%d).",
+            optionalTileHeight,
+            maxHeight);
+        error = buffer;
+        return false;
+      }
+    } catch (std::invalid_argument&) {
+      error = "Invalid web map service XML document";
+      return false;
+    }
+  } 
+
+  std::vector<std::string> configLayers{};
+  std::stringstream sstream(options.layers);
+  std::string layer;
+  const char delimiter = ',';
+  while (std::getline(sstream, layer, delimiter)) { 
+    if (layer.size() > 0) { 
+      configLayers.push_back(layer);
+    }
+  } 
+ 
+  tinyxml2::XMLElement* pOptionalServiceLayerLimit =
+      pService->FirstChildElement("LayerLimit");
+  if (pOptionalServiceLayerLimit) { 
+    try {
+      const int layerLimit = std::stoi(pOptionalServiceLayerLimit->GetText()); 
+      if (configLayers.size() > layerLimit) { 
+        char buffer[512];
+        std::sprintf(
+            buffer,
+            "the number of configured layers (%d) exceeds WMS LayerLimit %d",
+            static_cast<int>(configLayers.size()),
+            layerLimit);
+        error = buffer;
+        return false;
+      }
+    } catch (std::invalid_argument const&) { 
+      error = "Invalid web map service XML document";
+      return false;
+    }
+  } 
+
+  tinyxml2::XMLElement* pCapability = pRoot->FirstChildElement("Capability");
+  if (!pCapability) { 
+    error = "Invalid web map service XML document";
+    return false;
+  }
+  tinyxml2::XMLElement* pLayer = pCapability->FirstChildElement("Layer");
+  if (!pLayer) { 
+    error = "Invalid web map service XML document";
+    return false;
+  } 
+
+  std::vector<std::string> validLayerNames;
+  for (tinyxml2::XMLElement* pQueryableLayer = pLayer->FirstChildElement("Layer");
```

Comment:
WMS allows layers to be arbitrarily nested inside other layers, and any of these can be named layers that can be requested individually. So we can't assume all of the layers of interest are children of the root Layer. They need to be discovered recursively.

But really, it's probably ok to just pass through whatever layers the user requested straight through to the server. Let the server validate them and reject them if necessary, rather than trying to validate them manually against the GetCapabilities ourselves.